File: Back-END/todo-api/src/main/java/com/todo/api/service/TodoService.java
package com.todo.api.service;

import com.todo.api.model.Todo;
import com.todo.api.repository.TodoRepository;
import jakarta.persistence.EntityNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
public class TodoService {
    private final TodoRepository todoRepository;

    public List<Todo> getAllTodos() {
        return todoRepository.findAll();
    }

    public Todo getTodoById(Long id) {
        return todoRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Todo not found with id: " + id));
    }

    public List<Todo> getTodosByStatus(boolean completed) {
        return todoRepository.findByCompleted(completed);
    }

    public List<Todo> searchTodos(String title) {
        return todoRepository.findByTitleContainingIgnoreCase(title);
    }

    @Transactional
    public Todo createTodo(Todo todo) {
        return todoRepository.save(todo);
    }

    @Transactional
    public Todo updateTodo(Long id, Todo todoDetails) {
        Todo todo = getTodoById(id);
        todo.setTitle(todoDetails.getTitle());
        todo.setDescription(todoDetails.getDescription());
        todo.setCompleted(todoDetails.isCompleted());
        todo.setDueDate(todoDetails.getDueDate());
        return todoRepository.save(todo);
    }

    @Transactional
    public void deleteTodo(Long id) {
        Todo todo = getTodoById(id);
        todoRepository.delete(todo);
    }

    @Transactional
    public Todo toggleTodoStatus(Long id) {
        Todo todo = getTodoById(id);
        todo.setCompleted(!todo.isCompleted());
        return todoRepository.save(todo);
    }
}

File: Front-END/todo-app/src/services/TodoService.ts
import axios from 'axios';
import { Todo } from '../types/Todo';

const API_BASE_URL = 'http://localhost:8080/api/todos';

export const TodoService = {
    getAllTodos: async (): Promise<Todo[]> => {
        const response = await axios.get(API_BASE_URL);
        return response.data;
    },

    getTodoById: async (id: number): Promise<Todo> => {
        const response = await axios.get(`${API_BASE_URL}/${id}`);
        return response.data;
    },

    getTodosByStatus: async (completed: boolean): Promise<Todo[]> => {
        const response = await axios.get(`${API_BASE_URL}/status?completed=${completed}`);
        return response.data;
    },

    searchTodos: async (title: string): Promise<Todo[]> => {
        const response = await axios.get(`${API_BASE_URL}/search?title=${encodeURIComponent(title)}`);
        return response.data;
    },

    createTodo: async (todo: Omit<Todo, 'id'>): Promise<Todo> => {
        const response = await axios.post(API_BASE_URL, todo);
        return response.data;
    },

    updateTodo: async (id: number, todo: Todo): Promise<Todo> => {
        const response = await axios.put(`${API_BASE_URL}/${id}`, todo);
        return response.data;
    },

    deleteTodo: async (id: number): Promise<void> => {
        await axios.delete(`${API_BASE_URL}/${id}`);
    },

    toggleTodoStatus: async (id: number): Promise<Todo> => {
        const response = await axios.patch(`${API_BASE_URL}/${id}/toggle`);
        return response.data;
    }
};